Mathlib/Combinatorics/SimpleGraph/Matching.lean

```diff
@@ -226,4 +226,35 @@ lemma odd_matches_node_outside {u : Set V} {c : ConnectedComponent (Subgraph.del
 
 end Finite
 end ConnectedComponent
+
+/--
+  A graph is matching free if it has no perfect matching. It does not make much sense to
+  consider a graph being free of just matchtings, because any non-trivial graph has those.
+-/
+def IsMatchingFree (G : SimpleGraph V) := ∀ (M : Subgraph G), ¬Subgraph.IsPerfectMatching M
+
+lemma IsMatchingFree_mono {G G' : SimpleGraph V} (h : G ≤ G') (hmf : G'.IsMatchingFree) :
+    G.IsMatchingFree := by
+  intro x
+  by_contra! hc
+  apply hmf (x.map (SimpleGraph.Hom.ofLE h))
+  refine ⟨hc.1.map_ofLE h, ?_⟩
+  intro v
+  simp only [Subgraph.map_verts, Hom.coe_ofLE, id_eq, Set.image_id']
+  exact hc.2 v
+
+lemma exists_maximal_IsMatchingFree [Fintype V] [DecidableEq V]
+    (h : G.IsMatchingFree) : ∃ Gmax : SimpleGraph V,
+    G ≤ Gmax ∧ Gmax.IsMatchingFree ∧ ∀ G', G' > Gmax → ¬ G'.IsMatchingFree := by
```

Comment:
This is a general property of monotone predicates. Maybe Peter's #14721 will help here?

Reply:
I don't think it even needs monotinicity. It's more a specialization of generic facts to a Fintype. I'm not quite sure if that's an idiomatic statement. I was looking at Data.Fintype.Order for a landing place, but I see recently stuff in the Fintype namespace has been deprecated. Any thoughts on where this should go? 

As far as I understand Peter's work I don't think it's overlapping.

Reply:
> I was looking at Data.Fintype.Order for a landing place, but I see recently stuff in the Fintype namespace has been deprecated

Uh, what things were deprecated?

Reply:
[This stuff](https://github.com/leanprover-community/mathlib4/blob/604cb3610ee4c147dd3ef1966d0505c57cabbf0c/Mathlib/Data/Fintype/Order.lean#L215) (So mostly my question is, should this be here, and if yes should it be in the Finite or Fintype namespace?

(Currently working on following up on the comments below:))

Reply:
I commited something that I think makes some sense, but I'm very open to a better spot to put it.

Reply:
Oh, that's simply because some lemmas don't need constructive finiteness to be stated, so they shouldn't use constructive finiteness to be stated. Put yours in the namespace`Fintype` or `Finite` depending on whether they assume `Fintype` or `Finite`